azure/services/natgateways/spec_test.go

```diff
@@ -20,104 +20,129 @@ import (
+			name:         "create a new NAT Gateway spec when existing aso resource is nil",
+			spec:         fakeNatGatewaySpec,
+			existingSpec: nil,
+			expect: func(g *WithT, existing *asonetworkv1.NatGateway, parameters *asonetworkv1.NatGateway) {
+				g.Expect(parameters).NotTo(BeNil())
+				g.Expect(parameters).To(BeAssignableToTypeOf(&asonetworkv1.NatGateway{}))
+				g.Expect(parameters.Spec.Owner).NotTo(BeNil())
+				g.Expect(parameters.Spec.Owner.Name).To(Equal("my-rg"))
+				g.Expect(parameters.Spec.Location).NotTo(BeNil())
+				g.Expect(parameters.Spec.Location).To(Equal(locationPtr))
+				g.Expect(parameters.Spec.Sku.Name).NotTo(BeNil())
+				g.Expect(parameters.Spec.Sku.Name).To(Equal(standardSKUPtr))
+				g.Expect(parameters.Spec.PublicIpAddresses).To(HaveLen(1))
+				g.Expect(parameters.Spec.PublicIpAddresses[0].Reference.ARMID).To(Equal("/subscriptions/123/resourceGroups/my-rg/providers/Microsoft.Network/publicIPAddresses/my-natgateway-ip"))
+				g.Expect(parameters.Spec.Tags).NotTo(BeEmpty())
+				g.Expect(parameters.Spec.Tags).To(HaveKeyWithValue("Name", "my-natgateway"))
+				g.Expect(parameters.Spec.Tags).To(HaveKeyWithValue("sigs.k8s.io_cluster-api-provider-azure_cluster_my-cluster", "owned"))
 			},
-			expectedError: "",
 		},
 		{
-			name:     "get NatGateway when all values are present",
-			spec:     &fakeNatGatewaySpec,
-			existing: nil,
-			expect: func(g *WithT, result interface{}) {
-				g.Expect(result).To(BeAssignableToTypeOf(armnetwork.NatGateway{}))
-				g.Expect(result.(armnetwork.NatGateway).Location).To(Equal(ptr.To[string](fakeNatGatewaySpec.Location)))
-				g.Expect(result.(armnetwork.NatGateway).Name).To(Equal(ptr.To[string](fakeNatGatewaySpec.ResourceName())))
-				g.Expect(result.(armnetwork.NatGateway).Properties.PublicIPAddresses[0].ID).To(Equal(ptr.To[string](azure.PublicIPID(fakeNatGatewaySpec.SubscriptionID,
-					fakeNatGatewaySpec.ResourceGroupName(), fakeNatGatewaySpec.NatGatewayIP.Name))))
-				g.Expect(result.(armnetwork.NatGateway).Tags).To(Equal(fakeNatGatewaysTags))
+			name:         "create a NAT Gateway spec when there is an existing aso resource",
+			spec:         fakeNatGatewaySpec,
+			existingSpec: existingNatGateway,
+			expect: func(g *WithT, existing *asonetworkv1.NatGateway, parameters *asonetworkv1.NatGateway) {
+				diff := cmp.Diff(existing, parameters)
+				g.Expect(diff).To(BeEmpty())
 			},
-			expectedError: "",
 		},
 	}
-	for _, tc := range testCases {
+	format.MaxLength = 10000
+	for _, tc := range testcases {
 		tc := tc
 		t.Run(tc.name, func(t *testing.T) {
 			g := NewWithT(t)
 			t.Parallel()
 
-			result, err := tc.spec.Parameters(context.TODO(), tc.existing)
-			if tc.expectedError != "" {
-				g.Expect(err).To(HaveOccurred())
-				g.Expect(err).To(MatchError(tc.expectedError))
-			} else {
-				g.Expect(err).NotTo(HaveOccurred())
-			}
-			tc.expect(g, result)
+			result, _ := tc.spec.Parameters(context.TODO(), tc.existingSpec)
```

Comment:
This should be passed either a deep copy to mimic how Parameters is actually called if `existing` is being asserted on afterwards:
```suggestion
			result, _ := tc.spec.Parameters(context.TODO(), tc.existingSpec.DeepCopy())
```

Without a deep copy, Parameters ends up making `existing` and `parameters` point to the same struct which makes it impossible for the second test case to fail.